rustup-init.ps1

```diff
@@ -0,0 +1,174 @@
+#!/usr/bin/env powershell
+
+# This is just a little script that can be downloaded from the internet to
+# install rustup. It just does platform detection, downloads the installer and
+# runs it.
+
+#XXX: If you change anything here, please make the same changes in setup_mode.rs
+function Write-Usage {
```

Comment:
Warning - This is not idiomatic PowerShell.

PowerShell has a very rich, default help system, which means this is not required.  Instead should be using script parameters e.g.

```
#!/usr/bin/env powershell
<#
.Synopsis
   Installs RustUp
.DESCRIPTION
   This is just a little script that can be downloaded from the internet to install rustup. It just does platform detection, downloads the installer and runs it.
.PARAMETER DefaultHost
  The Host to install RustUp onto
.PARAMETER DefaultToolChain
  ...
  
.EXAMPLE
   ./rustup-init -DefaultHost '127.0.0.1'
#>

[CmdletBinding()]
Param(
  [Switch]$NoModiyPath,

  [Switch]$ShowVersion,

  [String]$DefaultHost,

  [ValidateSet("gnu", "msvc")]
  [String]$DefaultToolchain]
)

...
```

Reply:
Yeah, I went back and forth over this. The built in help system is much better (you get autocomplete and docs for free), however this was copying the strategy used in the `rustup-init.sh` implementation. In essence, it's as thin a wrapper around the `rustup-init.exe` executable as possible. Meaning, that if and when `rustup-init.exe` gains new flags and options, then this script passes those through without dealing with them directly. The strange situation here is that we're displaying a *copy* of the help usage from `rustup-init.exe --help` (I presume) so users know what options are available on the executable without downloading it first. If we went the param route, we'd have to wire each flag and option back into the executable and account for future changes. It's one of those situations where I'd have left a comment at the top to explain: "we know this isn't remotely idiomatic, but it's a pass-through to the executable". Does that make sense?

Reply:
Sure and I do understand the conundrum.  We have the same problem with our Puppet bootstrapping scripts being cross platform.

So perhaps there's a halfway solution here
1) For the options that we know conflict with PowerShell norms (specifically `-v`, `-h` and `-y`) they should be catered for and wired up appropriately.  We know that won't (99.99% sure) change.  Then you can use the Unbounded Parameters to capture the other options/commands etc. and pass verbatim.

This avoids the the churn when rustup adds new stuff.

2) You REALLY REALLY should have the documentation (`<# Synopsis...`) stuff at the top explaining that this script is a wrapper for xxxxx and copy the Write-Usage stuff into there.

Reply:
Ultimately it would be nice to autogenerate this script (and the .sh version too) from the rust-init build process but that's WAAAAY outside the scope of this PR.

Reply:
FWIW I agree with @glennsarti about the need to explain the intent in this file.

However it is not a wrapper - it is a bootstrapper; the difference is this: This script has to determine what exact precompiled rustup binary to download, then download it, and finally chain into it - after that it has no use and the user will never run it again. A wrapper we would expect to have the binary already present and be adapting it in some way, and be always used.

Does this matter? Maybe not, but if we are documenting the intent up the top for other people, I think we shouldn't confuse them with non-idiomatic use of such terms.